pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,348 @@
+		return false
+	}
+
+	// This can only happen if numAllowed has changed. If this happens flush all
+	// our current traces and don't register this request.
+	if numAllowed == 0 {
+		r.FlushAll()
+		return false
+	}
+
+	// Only increment the number of registered traces if the numRegistered
+	// hasn't changed. In the case of an ABA update, it does not break the
+	// invariant since some other trace was registered and deregistered, but
+	// there is still a slot available.
+	return r.numRegistered.CompareAndSwap(numRegistered, numRegistered+1)
+}
+
+func (r *RaftTracer) storeEntry(te kvserverpb.TracedEntry, baseCtx withContext) *traceValue {
+	tv := traceValue{
+		traced:  te,
+		baseCtx: baseCtx,
+	}
+	tv.ctx = logtags.AddTag(r.ctx, "id", redact.Safe(tv.String()))
+	tv.mu.appRespSent = make(map[raftpb.PeerID]bool)
+	tv.mu.storageAppendRespSent = make(map[raftpb.PeerID]bool)
+	r.m.Store(te.Index, &tv)
+	return &tv
+}
+
+// RegisterRemote is used to register a remote trace. This is used when we
+// receive a raft message over the wire with a request to continue tracing it.
+func (r *RaftTracer) RegisterRemote(te kvserverpb.TracedEntry) {
+	if !r.maybeRegister() {
+		return
+	}
+	// NB: We don't currently return remote traces, if we did, we would pass the
+	// remote ctx here and trace it. The problem is knowing when to send it
+	// back to the remote node.
+	tv := r.storeEntry(te, nil)
+	tv.logf(1, "registering remote trace %s", tv)
+}
+
+// withContext allows us to get the context from the object.
+type withContext interface {
+	Context() context.Context
+}
+
+// Register is called on an entry that is about to be proposed. This will begin
+// logging all subsequent updates to this entry.
+func (r *RaftTracer) Register(baseCtx withContext, ent raftpb.Entry) {
+	// Only register if there is a trace in the context and it is set to verbose
+	// logging.
+	span := tracing.SpanFromContext(baseCtx.Context())
+	if span == nil || span.RecordingType() != tracingpb.RecordingVerbose {
+		return
+	}
+
+	// If the index is nil, then we can't trace this entry. This can happen if
+	// there is a leader/leaseholder spilt. We don't have an easy way to handle
+	// this today, so don't attempt to trace it.
+	if ent.Index == 0 {
+		log.VEventf(baseCtx.Context(), 2, "skip registering raft proposal without index: %v", ent)
+		return
+	}
+
+	// This must be the last conditional. If this returns true we must call
+	// storeEntry to not leak a registered permit.
+	if !r.maybeRegister() {
+		log.VEvent(baseCtx.Context(), 2, "too many active raft traces, skipping")
+		return
+	}
+
+	// Grab the trace and span id now because the baseCtx.Context may be
+	// nil'ed later.
+	tv := r.storeEntry(
+		kvserverpb.TracedEntry{
+			Index:   kvpb.RaftIndex(ent.Index),
+			TraceID: span.TraceID(),
+			SpanID:  span.SpanID(),
+		},
+		baseCtx,
+	)
+	tv.logf(1, "registering local trace %s", tv)
+}
+
+// MaybeTrace will log the message if it is covered by a trace.
```

Comment:
nit: "will log" -> "logs". No much need in Future tense. Same in other comments.

Reply:
Done